File: internal/admin/ui/searchql.js
// JS mirror of internal/searchql/searchql.go. Exposes parseQuery, tokenize,
// and isUnindexedScan so the UI can render chips, toggle the unindexed-scan
// banner, and surface the OR-hint without a server round-trip. The shape of
// the returned AST matches the Go searchql.Query/Term fields case-for-case
// (field, value, wildcard, quoted, negated, headerName, statusFilter) so JS
// tests can mirror the Go parser tests.
(function () {
  var WILDCARD_NONE = 0;
  var WILDCARD_PREFIX = 1;
  var WILDCARD_SUBSTRING = 2;

  var FIELD_HOST = "host";
  var FIELD_PATH = "path";
  var FIELD_SERVICE = "service";
  var FIELD_BODY = "body";
  var FIELD_HEADERS = "headers";
  var FIELD_HEADER = "header";
  var FIELD_METHOD = "method";
  var FIELD_STATUS = "status";
  var FIELD_SOURCE_IP = "source_ip";
  var FIELD_CORRELATION_ID = "correlation_id";

  var KNOWN_FIELDS = {
    host: FIELD_HOST,
    path: FIELD_PATH,
    service: FIELD_SERVICE,
    body: FIELD_BODY,
    headers: FIELD_HEADERS,
    method: FIELD_METHOD,
    status: FIELD_STATUS,
    source_ip: FIELD_SOURCE_IP,
    correlation_id: FIELD_CORRELATION_ID,
  };

  var INDEXED_FIELDS = { host: true, path: true, service: true };
  var STRUCTURED_FIELDS = {
    method: true, status: true, source_ip: true, correlation_id: true,
  };
  var CANONICAL_METHODS = {
    GET: true, HEAD: true, POST: true, PUT: true, DELETE: true,
    CONNECT: true, OPTIONS: true, TRACE: true, PATCH: true,
  };

  function isSpace(c) {
    return c === " " || c === "\t" || c === "\n" || c === "\r" || c === "\v" || c === "\f";
  }

  function tokenize(s) {
    var tokens = [];
    var i = 0;
    var n = s.length;
    while (i < n) {
      while (i < n && isSpace(s[i])) i++;
      if (i >= n) break;
      var start = i;
      var inQuote = false;
      while (i < n) {
        var c = s[i];
        if (inQuote) {
          if (c === "\\" && i + 1 < n && s[i + 1] === "\"") { i += 2; continue; }
          if (c === "\"") { inQuote = false; i++; continue; }
          i++;
          continue;
        }
        if (c === "\"") { inQuote = true; i++; continue; }
        if (isSpace(c)) break;
        i++;
      }
      if (inQuote) {
        return { error: { token: s.slice(start), message: "unclosed quote" } };
      }
      tokens.push(s.slice(start, i));
    }
    return { tokens: tokens };
  }

  function canonicalHeaderName(name) {
    var out = "";
    var upper = true;
    for (var i = 0; i < name.length; i++) {
      var c = name.charCodeAt(i);
      if (upper) {
        if (c >= 97 && c <= 122) c -= 32;
        upper = false;
      } else if (c >= 65 && c <= 90) {
        c += 32;
      }
      var ch = String.fromCharCode(c);
      out += ch;
      if (ch === "-") upper = true;
    }
    return out;
  }

  function parseValue(valueRaw) {
    if (valueRaw.length > 0 && valueRaw.charAt(0) === "\"") {
      if (valueRaw.length < 2 || valueRaw.charAt(valueRaw.length - 1) !== "\"") {
        return { error: "unclosed quote" };
      }
      var inner = valueRaw.slice(1, -1);
      return { value: inner.replace(/\\"/g, "\""), wildcard: WILDCARD_NONE, quoted: true };
    }
    var v = valueRaw;
    var leading = v.length > 0 && v.charAt(0) === "*";
    if (leading) v = v.slice(1);
    var trailing = v.length > 0 && v.charAt(v.length - 1) === "*";
    if (trailing) v = v.slice(0, -1);
    if (v === "") {
      return { error: "value must contain a literal between wildcards" };
    }
    if (v.indexOf("*") !== -1) {
      return { error: "wildcards '*' are only allowed at the start or end of a value" };
    }
    var wildcard = WILDCARD_NONE;
    if (leading) wildcard = WILDCARD_SUBSTRING;
    else if (trailing) wildcard = WILDCARD_PREFIX;
    return { value: v, wildcard: wildcard, quoted: false };
  }

  function parseStatusValue(s) {
    if (s.length === 3 && s.charAt(1) === "x" && s.charAt(2) === "x") {
      var d = s.charCodeAt(0);
      if (d >= 49 && d <= 53) {
        return { class: s };
      }
    }
    if (!/^\d+$/.test(s)) {
      return { error: "must be an integer status code (e.g. 200) or class form (e.g. 2xx, 5xx)" };
    }
    var n = parseInt(s, 10);
    if (n < 100 || n > 599) {
      return { error: "must be an integer status code (e.g. 200) or class form (e.g. 2xx, 5xx)" };
    }
    return { exact: n };
  }

  function emptyTerm() {
    return {
      field: "",
      value: "",
      wildcard: WILDCARD_NONE,
      quoted: false,
      negated: false,
      headerName: "",
    };
  }

  function parseToken(raw) {
    var tok = raw;
    var negated = false;
    if (tok.length > 0 && tok.charAt(0) === "-") {
      negated = true;
      tok = tok.slice(1);
      if (tok === "") {
        return { error: { token: raw, message: "bare '-' is not a valid token" } };
      }
    }

    if (tok.length > 0 && tok.charAt(0) === "\"") {
      var pv = parseValue(tok);
      if (pv.error) return { error: { token: raw, message: pv.error } };
      var t = emptyTerm();
      t.value = pv.value; t.wildcard = pv.wildcard; t.quoted = pv.quoted; t.negated = negated;
      return { term: t };
    }

    var idx = tok.indexOf(":");
    if (idx < 0) {
      var pv2 = parseValue(tok);
      if (pv2.error) return { error: { token: raw, message: pv2.error } };
      var t2 = emptyTerm();
      t2.value = pv2.value; t2.wildcard = pv2.wildcard; t2.quoted = pv2.quoted; t2.negated = negated;
      return { term: t2 };
    }
    if (idx === 0) {
      return { error: { token: raw, message: "token must start with a key before ':'" } };
    }
    var key = tok.slice(0, idx);
    var valueRaw = tok.slice(idx + 1);
    if (valueRaw === "") {
      return { error: { token: raw, message: "empty value after ':'" } };
    }
    var keyLow = key.toLowerCase();
    if (!KNOWN_FIELDS[keyLow] && keyLow.indexOf("header.") === 0) {
      var name = key.slice("header.".length);
      if (name === "") {
        return { error: { token: raw, message: "empty header name after 'header.'" } };
      }
      if (name.indexOf("*") !== -1) {
        return { error: { token: raw, message: "wildcards are not supported in header name; use 'headers:' for fuzzy header search" } };
      }
      var pv3 = parseValue(valueRaw);
      if (pv3.error) return { error: { token: raw, message: pv3.error } };
      var t3 = emptyTerm();
      t3.field = FIELD_HEADER;
      t3.headerName = canonicalHeaderName(name);
      t3.value = pv3.value; t3.wildcard = pv3.wildcard; t3.quoted = pv3.quoted; t3.negated = negated;
      return { term: t3 };
    }
    var field = KNOWN_FIELDS[keyLow];
    if (!field) {
      return { error: { token: raw, message: "unknown key \"" + key + "\"" } };
    }
    var pv4 = parseValue(valueRaw);
    if (pv4.error) return { error: { token: raw, message: pv4.error } };
    if (STRUCTURED_FIELDS[field] && pv4.wildcard !== WILDCARD_NONE) {
      return { error: { token: raw, message: "wildcards are not supported on field \"" + key + "\"" } };
    }
    var term = emptyTerm();
    term.field = field;
    term.value = pv4.value;
    term.wildcard = pv4.wildcard;
    term.quoted = pv4.quoted;
    term.negated = negated;
    if (field === FIELD_METHOD) {
      var up = pv4.value.toUpperCase();
      if (!CANONICAL_METHODS[up]) {
        return { error: { token: raw, message: "unknown HTTP method \"" + pv4.value + "\"" } };
      }
      term.value = up;
    } else if (field === FIELD_STATUS) {
      var sf = parseStatusValue(pv4.value);
      if (sf.error) return { error: { token: raw, message: sf.error } };
      term.statusFilter = sf;
    }
    return { term: term };
  }

  function parseQuery(s) {
    var tk = tokenize(s);
    if (tk.error) return { terms: [], error: tk.error };
    var terms = [];
    for (var i = 0; i < tk.tokens.length; i++) {
      var pt = parseToken(tk.tokens[i]);
      if (pt.error) return { terms: [], error: pt.error };
      terms.push(pt.term);
    }
    return { terms: terms };
  }

  function isUnindexedScan(q) {
    if (!q || !q.terms) return false;
    for (var i = 0; i < q.terms.length; i++) {
      var t = q.terms[i];
      if (t.wildcard !== WILDCARD_SUBSTRING) continue;
      if (t.field === "" || INDEXED_FIELDS[t.field]) return true;
    }
    return false;
  }

  // chipDisplayValue mirrors the Go server's chipDisplayValue: status terms
  // use the original Exact/Class text; quoted values are wrapped in `"…"`
  // with `"` re-escaped; wildcarded values prepend or append `*`; bare values
  // pass through.
  function chipDisplayValue(term) {
    if (term.statusFilter) {
      if (term.statusFilter.exact) return String(term.statusFilter.exact);
      return term.statusFilter.class;
    }
    if (term.quoted) {
      return "\"" + String(term.value).replace(/"/g, "\\\"") + "\"";
    }
    if (term.wildcard === WILDCARD_PREFIX) return term.value + "*";
    if (term.wildcard === WILDCARD_SUBSTRING) return "*" + term.value + "*";
    return term.value;
  }

  function chipFromTerm(term) {
    var value = chipDisplayValue(term);
    var key = term.field;
    if (term.field === FIELD_HEADER) key = "header." + term.headerName;
    var freeform = term.field === "";
    var token = freeform ? value : key + ":" + value;
    if (term.negated) token = "-" + token;
    return {
      token: token,
      key: key,
      value: value,
      negated: !!term.negated,
      isHeader: term.field === FIELD_HEADERS || term.field === FIELD_HEADER,
      isFreeform: freeform,
    };
  }

  // shouldShowOrHint detects the literal `OR` between two field-qualified
  // terms — the three-consecutive-terms pattern from slice 06. Returns the
  // matched 0-based index of the middle `OR` token (so callers can highlight)
  // or -1 when the pattern is absent.
  function shouldShowOrHint(q) {
    if (!q || !q.terms || q.terms.length < 3) return -1;
    for (var i = 1; i < q.terms.length - 1; i++) {
      var mid = q.terms[i];
      if (mid.field !== "" || mid.value !== "OR" || mid.negated || mid.quoted) continue;
      if (mid.wildcard !== WILDCARD_NONE) continue;
      var prev = q.terms[i - 1];
      var next = q.terms[i + 1];
      if (prev.field !== "" && next.field !== "") return i;
    }
    return -1;
  }

  var api = {
    parseQuery: parseQuery,
    tokenize: tokenize,
    isUnindexedScan: isUnindexedScan,
    chipDisplayValue: chipDisplayValue,
    chipFromTerm: chipFromTerm,
    shouldShowOrHint: shouldShowOrHint,
    canonicalHeaderName: canonicalHeaderName,
    WildcardNone: WILDCARD_NONE,
    WildcardPrefix: WILDCARD_PREFIX,
    WildcardSubstring: WILDCARD_SUBSTRING,
    FieldHeader: FIELD_HEADER,
    FieldHeaders: FIELD_HEADERS,
  };

  if (typeof module !== "undefined" && module.exports) {
    module.exports = api;
  } else {
    (typeof window !== "undefined" ? window : globalThis).__searchql = api;
  }
})();

File: internal/searchql/searchql.go
// Package searchql parses the httpcatch search-language and compiles parsed
// queries into SQL fragments or in-memory predicates. The parser, AST, and
// compilers share a single Query type so the SQLite and memory backends behave
// identically.
//
// The accepted grammar is whitespace-separated tokens AND'd together. A token
// is either a key:value pair naming one of the known keys (host, path,
// service, body, method, status, source_ip, correlation_id, headers,
// header.<name>) or a bare freeform value that matches the term against a
// fixed union of fields ("Tier-1": host, path, service, all body and headers
// columns across both tables, plus correlated event request/response paths
// and bodies). Tokens may carry a leading '-' for negation, may quote values
// with double-quotes to preserve whitespace and treat '*' as literal, and may
// use '*' wildcards at the start, end, or both ends of an unquoted value.
package searchql

import (
	"fmt"
	"net/textproto"
	"strconv"
	"strings"
)

// Field is one of the recognised key names. The empty Field marks a freeform
// term whose value is matched against the Tier-1 union of fields.
type Field string

const (
	FieldHost          Field = "host"
	FieldPath          Field = "path"
	FieldService       Field = "service"
	FieldBody          Field = "body"
	FieldHeaders       Field = "headers"
	FieldHeader        Field = "header"
	FieldMethod        Field = "method"
	FieldStatus        Field = "status"
	FieldSourceIP      Field = "source_ip"
	FieldCorrelationID Field = "correlation_id"
)

// knownFields lists every key accepted by Parse. The `header.<name>:` form is
// not in this table — the parser detects the `header.` prefix separately so it
// can canonicalise the name segment and reject wildcards there.
var knownFields = map[string]Field{
	string(FieldHost):          FieldHost,
	string(FieldPath):          FieldPath,
	string(FieldService):       FieldService,
	string(FieldBody):          FieldBody,
	string(FieldHeaders):       FieldHeaders,
	string(FieldMethod):        FieldMethod,
	string(FieldStatus):        FieldStatus,
	string(FieldSourceIP):      FieldSourceIP,
	string(FieldCorrelationID): FieldCorrelationID,
}

// indexedFields are the fields whose default match is exact and whose leading
// wildcard forces a substring scan that bypasses the column index.
var indexedFields = map[Field]struct{}{
	FieldHost:    {},
	FieldPath:    {},
	FieldService: {},
}

// structuredFields are exact-match-only fields. Wildcards on these return a
// parse error.
var structuredFields = map[Field]struct{}{
	FieldMethod:        {},
	FieldStatus:        {},
	FieldSourceIP:      {},
	FieldCorrelationID: {},
}

// Wildcard classifies the wildcard shape of a term's value. Parse assigns it
// purely on syntax: a trailing-only '*' is WildcardPrefix; any leading '*'
// (with or without trailing) collapses to WildcardSubstring. Field semantics
// decide whether the wildcard actually changes the emitted predicate.
type Wildcard int

const (
	WildcardNone Wildcard = iota
	WildcardPrefix
	WildcardSubstring
)

// Term is one AND-leaf of a Query. Field is the parsed key; Value is the
// post-canonicalisation value with wildcard markers and surrounding quotes
// stripped. Wildcard is set from the value's syntactic shape; QuotedLiteral
// records whether the operator wrapped the value in double quotes (so '*'
// is literal); Negated is set when the token had a leading '-'.
type Term struct {
	Field         Field
	Value         string
	Wildcard      Wildcard
	QuotedLiteral bool
	Negated       bool

	// HeaderName is the canonical MIME header name (per
	// textproto.CanonicalMIMEHeaderKey) of a `header.<name>:` term. It is set
	// only when Field == FieldHeader.
	HeaderName string

	// StatusFilter is populated by Parse when Field == FieldStatus so both
	// compilers can dispatch on the same parsed form (exact vs class) without
	// re-parsing the value string.
	StatusFilter *StatusFilter
}

// StatusFilter holds a parsed status filter — either an exact code or a class
// (e.g. "2xx"). Exactly one of Exact and Class is set.
type StatusFilter struct {
	Exact int
	Class string
}

// ClassRange returns the inclusive [lo, hi] range covered by a class filter
// (e.g. "5xx" → 500, 599). Callers must only invoke this when Class is set.
func (sf *StatusFilter) ClassRange() (lo, hi int) {
	lo = int(sf.Class[0]-'0') * 100
	return lo, lo + 99
}

// Query is a flat list of terms AND'd together. An empty Query matches every
// record (no field constraint).
type Query struct {
	Terms []Term
}

// IsUnindexedScan reports whether any term forces an indexed dimension into
// substring matching — a leading wildcard against host, path, or service, or
// any leading-wildcard freeform term whose union expansion drives the indexed
// arms into substring matching. The caller uses this to surface the
// cost-class warning.
func (q Query) IsUnindexedScan() bool {
	for _, t := range q.Terms {
		if t.Wildcard != WildcardSubstring {
			continue
		}
		if t.Field == "" {
			return true
		}
		if _, ok := indexedFields[t.Field]; ok {
			return true
		}
	}
	return false
}

// ParseError carries the offending token text alongside the explanation. The
// admin layer reports the field as "q" and the message via Error().
type ParseError struct {
	Token   string
	Message string
}

func (e *ParseError) Error() string {
	if e.Token == "" {
		return e.Message
	}
	return fmt.Sprintf("%s: %s", e.Token, e.Message)
}

// Parse tokenises s (quote-aware, whitespace-separated) and produces a Query
// of AND'd terms. The grammar accepts only key:value tokens for the documented
// keys; everything else is rejected with a ParseError naming the offending
// token.
func Parse(s string) (Query, error) {
	tokens, err := tokenize(s)
	if err != nil {
		return Query{}, err
	}
	if len(tokens) == 0 {
		return Query{}, nil
	}

	terms := make([]Term, 0, len(tokens))
	for _, tok := range tokens {
		t, err := parseToken(tok)
		if err != nil {
			return Query{}, err
		}
		terms = append(terms, t)
	}
	return Query{Terms: terms}, nil
}

// tokenize walks s and emits one token per whitespace-separated run, treating
// the body of a double-quoted region as a single unit so internal whitespace
// survives the split. Escaped quotes (\") inside a quoted value are not
// treated as terminators. An unterminated quote returns a ParseError whose
// token is the remainder of the input from the opening quote.
func tokenize(s string) ([]string, error) {
	var tokens []string
	i := 0
	n := len(s)
	for i < n {
		for i < n && isASCIISpace(s[i]) {
			i++
		}
		if i >= n {
			break
		}
		start := i
		inQuote := false
		for i < n {
			c := s[i]
			if inQuote {
				if c == '\\' && i+1 < n && s[i+1] == '"' {
					i += 2
					continue
				}
				if c == '"' {
					inQuote = false
					i++
					continue
				}
				i++
				continue
			}
			if c == '"' {
				inQuote = true
				i++
				continue
			}
			if isASCIISpace(c) {
				break
			}
			i++
		}
		if inQuote {
			return nil, &ParseError{Token: s[start:], Message: "unclosed quote"}
		}
		tokens = append(tokens, s[start:i])
	}
	return tokens, nil
}

// isASCIISpace mirrors strings.Fields' notion of whitespace for our token
// stream — space, tab, newline, carriage-return, vertical-tab, form-feed.
func isASCIISpace(c byte) bool {
	switch c {
	case ' ', '\t', '\n', '\r', '\v', '\f':
		return true
	}
	return false
}

func parseToken(raw string) (Term, error) {
	tok := raw
	negated := false
	if strings.HasPrefix(tok, "-") {
		negated = true
		tok = tok[1:]
		if tok == "" {
			return Term{}, &ParseError{Token: raw, Message: "bare '-' is not a valid token"}
		}
	}

	// A token that begins with `"` is a fully quoted literal and never carries
	// a key prefix — a `:` inside the quotes is data, not the key separator.
	if strings.HasPrefix(tok, "\"") {
		value, wildcard, quoted, err := parseValue(tok)
		if err != nil {
			return Term{}, &ParseError{Token: raw, Message: err.Error()}
		}
		return Term{
			Value:         value,
			Wildcard:      wildcard,
			QuotedLiteral: quoted,
			Negated:       negated,
		}, nil
	}

	idx := strings.IndexByte(tok, ':')
	if idx < 0 {
		value, wildcard, quoted, err := parseValue(tok)
		if err != nil {
			return Term{}, &ParseError{Token: raw, Message: err.Error()}
		}
		return Term{
			Value:         value,
			Wildcard:      wildcard,
			QuotedLiteral: quoted,
			Negated:       negated,
		}, nil
	}
	if idx == 0 {
		return Term{}, &ParseError{Token: raw, Message: "token must start with a key before ':'"}
	}
	key := tok[:idx]
	valueRaw := tok[idx+1:]
	if valueRaw == "" {
		return Term{}, &ParseError{Token: raw, Message: "empty value after ':'"}
	}

	keyLow := strings.ToLower(key)
	field, ok := knownFields[keyLow]
	if !ok {
		if strings.HasPrefix(keyLow, "header.") {
			return parseHeaderTerm(raw, key, valueRaw, negated)
		}
		return Term{}, &ParseError{Token: raw, Message: fmt.Sprintf("unknown key %q", key)}
	}

	value, wildcard, quoted, err := parseValue(valueRaw)
	if err != nil {
		return Term{}, &ParseError{Token: raw, Message: err.Error()}
	}

	if _, isStructured := structuredFields[field]; isStructured && wildcard != WildcardNone {
		return Term{}, &ParseError{Token: raw, Message: fmt.Sprintf("wildcards are not supported on field %q", key)}
	}

	term := Term{
		Field:         field,
		Value:         value,
		Wildcard:      wildcard,
		QuotedLiteral: quoted,
		Negated:       negated,
	}

	switch field {
	case FieldMethod:
		upper := strings.ToUpper(value)
		if _, ok := canonicalHTTPMethods[upper]; !ok {
			return Term{}, &ParseError{Token: raw, Message: fmt.Sprintf("unknown HTTP method %q", value)}
		}
		term.Value = upper

	case FieldStatus:
		sf, err := parseStatusValue(value)
		if err != nil {
			return Term{}, &ParseError{Token: raw, Message: err.Error()}
		}
		term.StatusFilter = sf
	}

	return term, nil
}

// parseHeaderTerm handles `header.<name>:value` tokens. The name segment is
// canonicalised via textproto.CanonicalMIMEHeaderKey so equivalent operator
// spellings (User-Agent / user-agent / USER-AGENT) collapse into one AST. A
// wildcard in the name segment is rejected — fuzzy header-name search is the
// `headers:` keyword's job.
func parseHeaderTerm(raw, key, valueRaw string, negated bool) (Term, error) {
	name := key[len("header."):]
	if name == "" {
		return Term{}, &ParseError{Token: raw, Message: "empty header name after 'header.'"}
	}
	if strings.ContainsRune(name, '*') {
		return Term{}, &ParseError{Token: raw, Message: "wildcards are not supported in header name; use 'headers:' for fuzzy header search"}
	}
	value, wildcard, quoted, err := parseValue(valueRaw)
	if err != nil {
		return Term{}, &ParseError{Token: raw, Message: err.Error()}
	}
	return Term{
		Field:         FieldHeader,
		HeaderName:    textproto.CanonicalMIMEHeaderKey(name),
		Value:         value,
		Wildcard:      wildcard,
		QuotedLiteral: quoted,
		Negated:       negated,
	}, nil
}

// parseValue interprets the value portion of a token. Quoted values strip the
// outer double quotes and un-escape \" to "; their wildcard is always
// WildcardNone. Unquoted values may carry leading and/or trailing '*' — any
// leading '*' (with or without trailing) collapses to WildcardSubstring; a
// trailing-only '*' is WildcardPrefix; otherwise WildcardNone. An interior
// '*' or a value consisting entirely of wildcards is a parse error.
func parseValue(valueRaw string) (value string, wildcard Wildcard, quoted bool, err error) {
	if strings.HasPrefix(valueRaw, "\"") {
		if len(valueRaw) < 2 || !strings.HasSuffix(valueRaw, "\"") {
			return "", 0, false, fmt.Errorf("unclosed quote")
		}
		inner := valueRaw[1 : len(valueRaw)-1]
		value = strings.ReplaceAll(inner, "\\\"", "\"")
		return value, WildcardNone, true, nil
	}

	v := valueRaw
	leading := strings.HasPrefix(v, "*")
	if leading {
		v = v[1:]
	}
	trailing := strings.HasSuffix(v, "*") && len(v) > 0
	if trailing {
		v = v[:len(v)-1]
	}
	if v == "" {
		return "", 0, false, fmt.Errorf("value must contain a literal between wildcards")
	}
	if strings.Contains(v, "*") {
		return "", 0, false, fmt.Errorf("wildcards '*' are only allowed at the start or end of a value")
	}
	switch {
	case leading:
		return v, WildcardSubstring, false, nil
	case trailing:
		return v, WildcardPrefix, false, nil
	default:
		return v, WildcardNone, false, nil
	}
}

// canonicalHTTPMethods is the set of valid HTTP method names. Method values
// are stored uppercase; the parser normalises the input before validating.
var canonicalHTTPMethods = map[string]struct{}{
	"GET":     {},
	"HEAD":    {},
	"POST":    {},
	"PUT":     {},
	"DELETE":  {},
	"CONNECT": {},
	"OPTIONS": {},
	"TRACE":   {},
	"PATCH":   {},
}

// parseStatusValue accepts either an exact integer in [100, 599] or a class
// string of the form "Nxx" with N in [1, 5].
func parseStatusValue(s string) (*StatusFilter, error) {
	if len(s) == 3 && s[1] == 'x' && s[2] == 'x' {
		d := s[0]
		if d >= '1' && d <= '5' {
			return &StatusFilter{Class: s}, nil
		}
	}
	v, err := strconv.Atoi(s)
	if err != nil || v < 100 || v > 599 {
		return nil, fmt.Errorf("must be an integer status code (e.g. 200) or class form (e.g. 2xx, 5xx)")
	}
	return &StatusFilter{Exact: v}, nil
}
